### ISSUE TITLE
grpc / proto code gen

### ISSUE THREAD

**Issue by gedw99** [ent/ent#1473]
@rotemtam

Are the models and Services generated as different files ?

Example: https://github.com/amplify-edge/sys-share/tree/master/sys-account/proto/v2

It would be good for me if they are because i dont transport the protobuf over http only, but also over webrtc.

---

Also i saw the new Protobuf objects in a BLOB column. Looks really interesting. 
Are you planning to support Schema evolution at the DB using DB Schema Versioning, and hence GRPC Schema Evolution ( aka versioning ) ? Is this why your enabling this or is there another Use case for it ?

**Comment by rotemtam** [comment]
Hi @gedw99 ,
Thanks for reaching out!

As you can see in https://github.com/ent/contrib/tree/master/entproto/internal/todo/ent/proto/entpb, the proto generation code yields a single .proto file per package. I'm not sure I understand why separating the .proto file into two chunks would be useful in your use case, can you elaborate? In any case, the models are generating in addition to the service definition so you can just not use the service def if you don't need it for a specific use case. 

With regards to storing binary pb in a BLOB column, what is currently supported is a very naive use case - we just show how to use `ValueScanner` to help ent figure out how to serialize/deserialize the bytes into a Go struct. It is related  to schema versioning in the sense that it is stored in an ent field, which will be versioned. If the proto message is a product of entproto generation, then the proto definition will also be subject to versioning (simply because its a direct, deterministic product of it) , if that's what you mean.

**Comment by gedw99** [comment]
Everything you said about GRPC schema evolution makes perfect sense. Is there going to be a integration of transaction hook into this GRPC approach So I can do event updates and hence manage quasi materialised views and cdc patterns ? The Ent hooks layer looks really useful for managing views in the dB and then also change events on those views.

The reason for separation in our case is to do with a fair bit of post processing . We generate a cli off the proto and also a data boot strapped that is integrated with the generated cli . It makes it easy to play with a system then . Ask me if you need more info.

**Comment by rotemtam** [comment]
Good questions @gedw99 !

Out of the box you can use [runtime hooks](https://entgo.io/docs/hooks#runtime-hooks) since those can be configured on the ent client before it is injected into the generated gRPC service. In the future, we may add a way to integrate with transaction hooks. How would you expect such an API to look like? 

A general word of advice, depending on your requirements, but writing to a "CDC" log in the application layer, even if only after a transaction was committed is unsafe - a crash/process restart that happens after the commit but before the CDC "send" was complete, can leave your application in an inconsistent state. What I've seen in the past is the usage of infrastructure like Kafka Connect to pull from the _actual_ database CDC into durable storage (such as Kafka) and process it from there.

**Comment by gedw99** [comment]
Yep transaction hooks are useful for general things like then sending an event to your event bus to tell another system something happened. Like when you store files in minio, but model the directory structure in the SQL DB for example. 
A CDC hook tells you the before and after state typically. But these hooks are not for the cdc use case, so i think maybe the schema, table and row data is enough, but i am only guessing.

About CDC and linearizability -  100% agree that using hooks for CDC is unsafe for reasons you raised.
And yep kafka connect and https://debezium.io/ are the usual suspects.

There are IMHO better ones emerging though.
NATS / LIftbridge ( both golang ) with golang cdc tools such as:

cockroach
https://github.com/bobvawter/cacheroach
- uses https://www.cockroachlabs.com/docs/stable/changefeed-for.html, which is NOT enterprise only.

cockroach enterprise
https://github.com/cockroachdb/cdc-sink

postresql
https://github.com/lawrencejones/pgsink
- uses standard pg

sqlite
https://github.com/benbjohnson/litestream
- close to being able to do cdc


I guess this all leads back to how you all see graphql subscriptions being supported ?
1. Hooks to a Data Loader 
Or
2. Materialised views with CDC so you can push the change to the client.
Or 

I can image also the plan is to support GRPC and Graphql.
Large project uses grpc and then allow the graphql "federation" ( which is a loaded work i know) to use the grpc layer to tell the graphql layer about a change. That change being stored in Kafka or NATS or somewhere and being originally generated from CDC or Hooks... 

Hope i am not raising things that are out of scope.. I am just quite eager about what ENT represents and can see how potent it can be for golang projects needing real time systems.

**Comment by kkom** [comment]
The CDC discussion is very interesting and relevant to one of the reasons why we're interested in exploring entproto. Thanks for raising it!

@rotemtam - I think that using the transactional outbox pattern could make CDC in the application layer safe. Am I missing something? Or do you think there is any reason why it would be difficult to implement/ not a great idea with entproto?

**Comment by gedw99** [comment]
@kkom here is an implementation of the  transactional outbox pattern in golang for postgresql.

https://github.com/obsidiandynamics/goharvest/wiki/Comparison-of-messaging-patterns

**Comment by rotemtam** [comment]
@kkom yes, should be quite possible to do with ent generically. If you or @gedw99 (or both!) want to take a stab at it, sync with me on our slack channel so we can coordinate.

**Comment by kkom** [comment]
Great! Thanks for the quick response both!

To set the right expectations - definitely not now, but maybe in a few months time?

It depends on whether ent + entproto is something we want to seriously explore as the fundamental data layer in the company I'm at, similar to https://engineering.fb.com/2013/06/25/core-data/tao-the-power-of-the-graph/

It's a much wider discussion so I'll start a new thread in the discussions forum :)

**Comment by gedw99** [comment]
Yep would be great to collaborate on this.

**Comment by gedw99** [comment]
Please post forum link

**Comment by gedw99** [comment]
i am currently playing with using https://github.com/MaterializeInc/materialize btw for CDC. 

Its rust, not golang.

It supports many sources: https://materialize.com/docs/sql/create-source/
postresql, s3, files.

The sink is the Materialised Views exposed by Materialize, and the change stream off those Views.

**Comment by wenerme** [comment]
I really think ent should not go too deep on cdc, orm is not database, what ent can get is not always right, ent provide hook to impl basic cdc is good enough, by streaming to nats can do a lot of cool things, if 100% cdc is required, should look at the underlying db.

**Comment by giautm** [comment]
In CDC architecture, you should not put them in a single application. Why?
- If your application crash, events may not publish to the stream.
- If your wait publishes succession then commit the transaction, it may slow down your application.
See [the blog from Nginx](https://www.nginx.com/blog/event-driven-data-management-microservices/) for atomicity publish.

You can use 3rd tools like Debezium, Kafka Connect to capture data change from your database to achieving atomicity.

**Comment by gedw99** [comment]
The ngoni article says you need an event table if you want atomicity

**Comment by gedw99** [comment]
hey @rotemtam and @giautm 

Has anyone moved forward on using the pattern that @giautm suggested ? That being the Nginx atomicity publish approach of using an event table in the DB.

I am using Ent for a project and need to publish changes in the DB out to a Search system based on Bluge ( here it is https://github.com/mosuka/phalanx ).
